### ISSUE TITLE
Anomaly error message when compiling an inductive definition

### ISSUE THREAD

**Issue by ianshil** [coq/coq#13300]
Hi all, 

First, know that I am using CoqIDE 8.12.0 and that I am rather new to Coq.

Second, here is my problem. I am trying to use some definitions from a file called dd_fc.v which was developed by another researcher in my lab using an earlier version of CoqIDE. When I try to compile it, the following error message appears when the compiler hits an inductive definition called botRule_fc:

> Anomaly
> "File "kernel/univ.ml", line 879, characters 4-10: Assertion failed."
> Please report at http://coq.inria.fr/bugs/.

Note that for the definitions derrec_concl, derrec_fc and dersrec_fc, appearing before botRule_fc, the compiler warns me with the following message:

> Not a truly recursive fixpoint. [non-recursive,fixpoints]

For some reason this last message does not appear on CoqIDE 8.11, even though I am not sure if this is related. 

Here is the code required to see the phenomenon I am describing happening: 

```coq
Set Implicit Arguments.
Notation "[ ]" := nil (format "[ ]") .
Notation "[ x ]" := (cons x nil) .
Notation "[ x ; y ; .. ; z ]" :=  (cons x (cons y .. (cons z nil) ..)) .
Open Scope list_scope.

Axiom map : forall {A B}, (A -> B) -> list A -> list B.

Polymorphic Definition rlsT W := list W -> W -> Type.

Inductive derrec X (rules : list X -> X -> Type) (prems : X -> Type) :
  X -> Type :=
  | dpI : forall concl,
    prems concl -> derrec rules prems concl
  | derI : forall ps concl,
    rules ps concl -> dersrec rules prems ps -> derrec rules prems concl
with dersrec X (rules : list X -> X -> Type) (prems : X -> Type) :
  list X -> Type :=
  | dlNil : dersrec rules prems []
  | dlCons : forall seq seqs,
    derrec rules prems seq -> dersrec rules prems seqs ->
    dersrec rules prems (seq :: seqs)
    .

Inductive derrec_fc X rules (prems : X -> Type) : Type :=
  | fcI : forall concl, derrec rules prems concl -> derrec_fc rules prems.

Inductive dersrec_fcs X rules (prems : X -> Type) : Type :=
  | fcsI : forall concls, dersrec rules prems concls -> dersrec_fcs rules prems.

Fixpoint derrec_concl X rules prems concl
  (der : @derrec X rules prems concl) :=
  match der with
    | dpI _ _ c _ => c
    | derI c _ _ => c
  end.

Fixpoint derrec_fc_concl X rules prems
  (der : @derrec_fc X rules prems) :=
  match der with
    | fcI d => derrec_concl d
  end.

Fixpoint dersrec_concls X rules prems concls
  (ders : @dersrec X rules prems concls) :=
  match ders with
    | dlNil _ _ => []
    | dlCons d ds => derrec_concl d :: dersrec_concls ds
  end.


Fixpoint dersrec_fc_concls X rules prems
  (ders : @dersrec_fcs X rules prems) :=
  match ders with
    | fcsI ds => dersrec_concls ds
  end.

Fixpoint dersrec_trees X rules prems concls
  (ders : @dersrec X rules prems concls) :=
  match ders with
    | dlNil _ _ => []
    | dlCons d ds => fcI d :: dersrec_trees ds
  end.

Definition nextup W rules prems (dt : @derrec_fc W rules prems) :=
  match dt with
    | fcI (dpI _ _ _ _) => []
    | fcI (derI _ _ ds) => dersrec_trees ds
  end.

Inductive botRule_fc X rules prems (der : @derrec_fc X rules prems) : rlsT X :=
  | botRule_fcI : botRule_fc der
    (map (@derrec_fc_concl _ _ _) (nextup der)) (derrec_fc_concl der).
```

Can someone tell me what is happening here? 

Thanks in advance. 

Best,
Ian

**Comment by SkySkimmer** [comment]
Reduced example
~~~coq
Polymorphic Definition type := Type.

Inductive bad : type := .
~~~
Looks like we're too aggressive in making the universe Prop.

**Comment by Zimmi48** [comment]
@ianshil: Thanks for your report. FYI, the new warning you see (Not a truly recursive fixpoint) is not related with the bug. It is a new warning introduced in 8.12 that is raised when you use `Fixpoint` but should use `Definition` instead.

**Comment by Zimmi48** [comment]
@SkySkimmer Any idea when this bug was introduced and whether we should consider it high priority and find someone to fix it?

**Comment by SkySkimmer** [comment]
Could be as early as d9ac4c22a3a6543959d413120304e356d625c0f9
I'll look into a fix.

**Comment by Zimmi48** [comment]
If it were an old bug, what would have make it surface in 8.12?

**Comment by SkySkimmer** [comment]
Right, 2ed097cfba9136a0fba9b961d24c408077fac11d is the first bad commit
Before the lbound stuff we would be stopped from lowering by the `Evd.check_leq evd Univ.type0_univ duu`

**Comment by Zimmi48** [comment]
When this bug was discussed during the Coq Call, people said that there is an easy workaround (adding an explicit `Type` annotation?). Could the people that know what this workaround is let @ianshil know, since the bug is probably not going to be fixed in 8.12.1?
